### DISCUSSION TITLE
Interactivity API should have a method to access root context when using wp-each directive

### DISCUSSION THREAD

**Discussion by sethrubenstein** [WordPress/gutenberg#65770] [Interactivity API]
What I'm proposing is a new function in the api, `getRootContext`. This would return the context of the current elements root element, so one interactive element up. 

Take this example of using `wp-each` directive: 

```html
<ul data-wp-context='{ "uniqueUrlSlug": "xyz", "list": [ "hello", "hola", "olá" ] }'>
  <template data-wp-each="context.list" >
    <li data-wp-text="context.item">
      <button data-wp--click="actions.onClick">Go to item</button>
    </li>
  </template>
</ul>
```

And consider this view.js

```js
/**
 * WordPress Dependencies
 */
import { store, getContext, getElement } from '@wordpress/interactivity';

const { state, actions } = store('my-namespace/my-block', {
  actions: {
    onClick: () => {
      const context = getContext();
      // Create a new url using context from my room element, uniqueUrlSlug
      // When we examine context, it does not contain uniqueUrlSlug because the context is now local to the template `context.item` so only the list items would appear in context. 
    }
  }
});
```

What if there was a `getRootContext` function, that would return the context of the root element? This would be useful for more than just in wp-each, I could see myself using this to get the content of the interactive namespace one element up, and so on (without getting too ridiculous). This would allow a little more broader use of context vs global state. Ideally you would also be able to set the root element's context as well not just read.

**Comment by gziolo** [comment]
Out of curiosity, what’s the reason you can’t access `context.uniqueUrlSlug` from your example included? Why would `data-wp-each` change anything with how the context works? Before introducing a new API method that is going to cause confusion when used outside of `data-wp-each`, it would be good to revisit whether the current limitation can be removed and the way context works potentially unified across all child HTML elements.

**Reply by sethrubenstein** [reply]
Hi @gziolo. 

I'm referencing use cases like a dropdown component like this where the root element has context and then when using getContext in template items only returns and sets context for that item with no ability to get or set the context of the element the template is getting data from.

Render Markup:
```php
<?php
$block_wrapper_attrs = get_block_wrapper_attributes([
	'id' => wp_unique_id('prc-platform__dropdown-list-'),
	'data-wp-interactive' => wp_json_encode([
		'namespace' => 'prc-platform/dropdown-list',
	]),
	'data-wp-context' => wp_json_encode([
		'expanded' => false,
		'value' => '',
		'choices' => [
			[
				'label' => 'Option 1',
				'value' => 'option-1',
			],
			[
				'label' => 'Option 2',
				'value' => 'option-2',
			]
		],
	]),
	'data-wp-class--is-open' => 'context.expanded',
]);
$options_template = wp_sprintf(
	'<template data-wp-each--choice="%s" data-wp-each-key="context.choice.value"><li><button data-wp-text="context.choice.label" data-wp-on--click="actions.onClick" type="button"></button></li></template>',
	'context.choices',
);
echo wp_sprintf(
	'<div %s><div class="open-close-toggle" data-wp-on--click="actions.onExpand">Expand</div><ul role="listbox">%s</ul></div>',
	$block_wrapper_attrs,
	$options_template,
);
```

view.js Markup
```js
/**
 * WordPress Dependencies
 */
import {
	store,
	getContext,
} from '@wordpress/interactivity';

const { state, actions } = store('prc-platform/dropdown-list', {
	actions: {
		onClick: () => {
			const context = getContext();
			console.log(context);
			// All we get is context.choice
			// This contains { label, value }
			// But it does not contain {expanded, value} from the root element
			// So, doing something like this, will not work:	
			context.expanded = false;
			context.value = context.choice.value;
			// Ideally, getContext() would return and set accoridngly { choice: {value, label}, expanded, value} 
		}
	}
});
```

**Reply by gziolo** [reply]
That's exactly how I understood it. Thank you for providing more details. @DAreRodz or @michalczaplinski, do you know why there are limitations to the context exposed through the `each` directive? Does it need to be special or was it missed to propagate the entire context?

**Reply by michalczaplinski** [reply]
I believe that @DAreRodz and/or @luisherranz decided on this at some point. 

I don't know the reason, but I agree that the behavior described by @sethrubenstein is what I'd also expect. I'd also expect `getContext()` to return the local value inside the `wp-each` AND the parent context.

**Reply by sirreal** [reply]
Merging the contexts (`each` and whatever other context is present) is exactly what I'd expect. It's surprising to me that isn't the case now, if I'm understanding correctly.

**Reply by DAreRodz** [reply]
:wave:  As explained in [Working with local context](https://developer.wordpress.org/block-editor/reference-guides/interactivity-api/core-concepts/undestanding-global-state-local-context-and-derived-state/#local-context:~:text=Nesting%20local%20contexts), part of the Interactivity API's [Core Concepts](https://developer.wordpress.org/block-editor/reference-guides/interactivity-api/core-concepts/) guide:

> Local contexts can be nested, with child contexts inheriting and potentially overriding values from parent contexts.

The `data-wp-each` directive effectively acts like `data-wp-context`, creating a new context that, in this case, also includes a new property (`item`) with the specific list item. That means properties from parent contexts should be inherited.

---

The code snippet below is based on the opening post example. I tested it and seems to work as expected. 🙂

It can be easily tested by pasting its content into a Custom HTML block in a page/post/template with interactive blocks (so the Interactivity API is enqueued automatically).

I've made a couple of changes, like adding the `data-wp-interactive` directive with the appropriate namespace and moving the `data-wp-text` directive to a `<span>` element, in order not to overwrite the `<li>` content and get rid of the `<button>` element.

```html
<ul
  data-wp-interactive="my-namespace/my-block"
  data-wp-context='{ "uniqueUrlSlug": "xyz", "list": [ "hello", "hola", "olá" ] }'
>
  <template data-wp-each="context.list" >
    <li>
      <span data-wp-text="context.item"></span>
      <button data-wp-on--click="actions.onClick">Go to item</button>
    </li>
  </template>
</ul>

<script type="module">
import { store, getContext, getElement } from '@wordpress/interactivity';

const { state, actions } = store('my-namespace/my-block', {
  actions: {
    onClick: () => {
      const context = getContext();
      console.log(`${context.uniqueUrlSlug}/${context.item}`);
    }
  }
});
</script>
```

**Reply by DAreRodz** [reply]
@sethrubenstein, regarding the code example you mention in https://github.com/WordPress/gutenberg/discussions/65770#discussioncomment-10881791, could it be that the browser dev tools do not show the inherited properties even though they exist?

The inheritance mechanism works with proxies. As you can see in the image below, the context object only contains `choice`, but the `expanded`, `value`, and `choices` properties are accessible too. 🙂 

![Screenshot 2024-10-15 at 13 24 52](https://github.com/user-attachments/assets/4b6d8669-72a3-43c3-bcdd-282dace689ce)

**Reply by sethrubenstein** [reply]
@DAreRodz That is exactly it. All this time and I didn't think I could access the root values because the console wasn't showing them 🤦‍♂️. Just tested this and I can access and set these values.

**Reply by sirreal** [reply]
Debugging the proxied values is tricky. I often use that spread trick.

It would be good to document this behavior at least. Are there things we could do to make it easier to work with?
